File: Java/프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.java
import java.util.*;
class Solution {
    public int solution(int[][] board, int[] moves) {
        int answer = 0;
        int[] existRowIndex = new int[board.length];
        ArrayList<Integer> result = new ArrayList<>();
        //init
        for(int i=0; i<existRowIndex.length; i++) {
            existRowIndex[i] = -1;
        }
        for(int i=0; i<board.length; i++) {
            for(int j=0; j<board[0].length; j++) {
                if(board[i][j] != 0) {
                    if(existRowIndex[j] == -1){
                        existRowIndex[j] = i;
                        // answer+=i; //확인용 {3,2,1,3,1}
                    }
                }
            }
        }
        
        for(int i=0; i<moves.length; i++) {
            if(existRowIndex[moves[i]-1] < board.length && existRowIndex[moves[i]-1] != -1) {
                result.add(board[existRowIndex[moves[i]-1]][moves[i]-1]);
                if(result.size()>1 && result.get(result.size()-1) ==result.get(result.size()-2)) {
                    result.remove(result.size()-1);
                    result.remove(result.size()-1);
                    answer+=2;
                }
                existRowIndex[moves[i]-1]++;
            }
        }
        
        
        return answer;
    }
}

File: Kotlin/프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.kt
class Solution {
    fun solution(board: Array<IntArray>, moves: IntArray): Int {
        var answer = 0
        var lastDollIndex = Array(board[0].size) {-1}
        var pickedDolls = mutableListOf<Int>()
        for(i in 0..board.lastIndex) {
            for(j in 0..board[i].lastIndex) {
                if(board[i][j] != 0 && lastDollIndex[j] == -1) {
                    lastDollIndex[j] = i
                }
            }
        }
        moves.forEach{
            if(lastDollIndex[it-1] != -1 && lastDollIndex[it-1] < board.size) {
                pickedDolls.add(board[lastDollIndex[it-1]][it-1])
                lastDollIndex[it-1] +=1
                if(pickedDolls.size > 1 && pickedDolls[pickedDolls.lastIndex] 
                   == pickedDolls[pickedDolls.lastIndex-1]) {
                    pickedDolls.removeAt(pickedDolls.lastIndex)
                    pickedDolls.removeAt(pickedDolls.lastIndex)
                    answer+=2
                    } 
                }
            }
        return answer
    }
}